### PR TITLE
AWS Event Stream Protocol Support

### PR DESCRIPTION
**Pull request by skmcgrail**
Adds support for AWS Event Streaming to the following clients:
* `service/transcribestreaming`:
  * StartStreamTranscription
  * StartMedicalStreamTranscription
* `service/kinesis`:
   * SubscribeToShard
* `service/s3`:
  * SelectObjectContent
* `service/lexruntimev2`:
  *  StartConversation

Requires https://github.com/aws/smithy-go/pull/331

Resolves https://github.com/aws/aws-sdk-go-v2/issues/1234
Resolves https://github.com/aws/aws-sdk-go-v2/issues/174

### CODE REVIEW

File: internal/protocoltest/awsrestjson/types/types.go [hunk 1]
```diff
@@ -16,27 +16,27 @@ type ComplexNestedErrorData struct {
...
-//  MyUnionMemberBooleanValue
-//  MyUnionMemberNumberValue
 //  MyUnionMemberBlobValue
-//  MyUnionMemberTimestampValue
+//  MyUnionMemberBooleanValue
 //  MyUnionMemberEnumValue
 //  MyUnionMemberListValue
 //  MyUnionMemberMapValue
-//  MyUnionMemberStructureValue
+//  MyUnionMemberNumberValue
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply internal/protocoltest/awsrestjson/types/types.go:24]: This was changed to collect them up into a `TreeSet` before filtering our any of the structure members that have an `ErrorTrait` on them for event streams. So this is what would have caused the sort order to change on these.

File: aws/middleware/metadata.go [hunk 2]
```diff
@@ -165,3 +165,16 @@ func GetEndpointSource(ctx context.Context) (v aws.EndpointSource) {
...
+// GetSigningCredentials returns the credentials that were used for signing if set on context.
+func GetSigningCredentials(ctx context.Context) (v aws.Credentials) {
+	v, _ = middleware.GetStackValue(ctx, signingCredentialsKey{}).(aws.Credentials)
+	return v
+}
+
+// SetSigningCredentials sets the credentails used for signing on the context.
+func SetSigningCredentials(ctx context.Context, value aws.Credentials) context.Context {
+	return middleware.WithStackValue(ctx, signingCredentialsKey{}, value)
+}
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply aws/middleware/metadata.go:180]: Let's sync offline on this topic.

File: service/transcribestreaming/internal/testing/testing.go [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Package testing is used for custom unit tests for transcribestreaming.
+package testing
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply service/transcribestreaming/internal/testing/testing.go:2]: Correct

File: internal/protocoltest/restxml/types/types.go [hunk 4]
```diff
@@ -99,28 +99,20 @@ type XmlNestedUnionStruct struct {
...
+//  XmlUnionShapeMemberFloatValue
 //  XmlUnionShapeMemberIntegerValue
 //  XmlUnionShapeMemberLongValue
-//  XmlUnionShapeMemberFloatValue
-//  XmlUnionShapeMemberDoubleValue
-//  XmlUnionShapeMemberUnionValue
+//  XmlUnionShapeMemberShortValue
+//  XmlUnionShapeMemberStringValue
 //  XmlUnionShapeMemberStructValue
+//  XmlUnionShapeMemberUnionValue
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply internal/protocoltest/restxml/types/types.go:111]: See earlier comment reply.

File: service/transcribestreaming/protocol_test.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Code generated by smithy-go-codegen DO NOT EDIT.
+
+package transcribestreaming
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply service/transcribestreaming/protocol_test.go:3]: These have always been generated for all clients and is not a change in behavior.

File: service/transcribestreaming/eventstream.go [hunk 6]
```diff
@@ -0,0 +1,788 @@
...
+	return w
+
+}
+
+func (w *audioStream) Send(ctx context.Context, event types.AudioStream) error {
+	return w.send(ctx, event)
+}
+
+func (w *audioStream) send(ctx context.Context, event types.AudioStream) error {
+	if err := w.err.Err(); err != nil {
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply service/transcribestreaming/eventstream.go:109]: Technically not for HTTP binding protocols, but for RPC protocols the two signatures are different. The codegen is less complicated if the pattern is consistent across bindings and RPC. Go compiler **should** inline these anyways.

File: aws/protocol/eventstream/decode.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
+package eventstream
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply aws/protocol/eventstream/decode.go:1]: Discussed offline.

**Review by jasdel — APPROVED** [review]

**Pull request merged**